### ISSUE TITLE
modules are not working

### ISSUE THREAD

**Issue by TheEric7** [vmware/ansible-module-vcloud-director#167]
_---
- name: vCloudDirectorAnsible
  hosts: localhost
  environment:
    env_user: my_user
    env_password: my_pass'
    env_host: 10.10.9.200
    env_org: system
    env_api_version: 30.0
    env_verify_ssl_certs: false
  tasks:
  - name: create org
    vcd_org:
      org_name: acme_sample
      full_name: acme_sample
      is_enabled: true
      state: present
    register: output

  - name: create org output
    debug:
      msg: '{{ outpu }}'

When I run it gives me the following error:

[srvc_ansible@ansible-srv]$ ansible-playbook main.yml

[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [vCloudDirectorAnsible] ********************************************************************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************************************************************************
ok: [localhost]

TASK [create org] *******************************************************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: Value of unknown type: <class 'pyvcloud.vcd.exceptions.UnauthorizedException'>, Status code: 401, <empty response body> (request id: 87a157ab-ca00-4f5a-95e0-e4f15a0b9bd5)
fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_vcd_org_payload_22w1ajqf/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py\", line 327, in main\n  File \"/tmp/ansible_vcd_org_payload_22w1ajqf/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py\", line 131, in manage_states\n  File \"/tmp/ansible_vcd_org_payload_22w1ajqf/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py\", line 173, in create\n  File \"/tmp/ansible_vcd_org_payload_22w1ajqf/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py\", line 161, in get_org\n  File \"/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py\", line 1562, in get_org_by_name\n    orgs = self._get_wk_resource(_WellKnownEndpoint.ORG_LIST)\n  File \"/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py\", line 1659, in _get_wk_resource\n    return self.get_resource(self._get_wk_endpoint(wk_type))\n  File \"/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py\", line 1452, in get_resource\n    extra_headers=extra_headers)\n  File \"/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py\", line 1169, in _do_request\n    _objectify_response(response, objectify_results))\n  File \"/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py\", line 1177, in _response_code_to_exception\n    raise UnauthorizedException(sc, request_id, objectify_response)\npyvcloud.vcd.exceptions.UnauthorizedException: Status code: 401, <empty response body> (request id: 87a157ab-ca00-4f5a-95e0-e4f15a0b9bd5)\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/home/srvc_ansible/.ansible/tmp/ansible-tmp-1613711762.2601597-16265-72118068949560/AnsiballZ_vcd_org.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/home/srvc_ansible/.ansible/tmp/ansible-tmp-1613711762.2601597-16265-72118068949560/AnsiballZ_vcd_org.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/srvc_ansible/.ansible/tmp/ansible-tmp-1613711762.2601597-16265-72118068949560/AnsiballZ_vcd_org.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible.modules.vcd_org', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib64/python3.6/runpy.py\", line 205, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.6/runpy.py\", line 96, in _run_module_code\n    mod_name, mod_spec, pkg_name, script_name)\n  File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_vcd_org_payload_22w1ajqf/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py\", line 341, in <module>\n  File \"/tmp/ansible_vcd_org_payload_22w1ajqf/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py\", line 335, in main\n  File \"/tmp/ansible_vcd_org_payload_22w1ajqf/ansible_vcd_org_payload.zip/ansible/module_utils/basic.py\", line 2214, in fail_json\n  File \"/tmp/ansible_vcd_org_payload_22w1ajqf/ansible_vcd_org_payload.zip/ansible/module_utils/basic.py\", line 2187, in _return_formatted\n  File \"/tmp/ansible_vcd_org_payload_22w1ajqf/ansible_vcd_org_payload.zip/ansible/module_utils/basic.py\", line 420, in remove_values\n  File \"/tmp/ansible_vcd_org_payload_22w1ajqf/ansible_vcd_org_payload.zip/ansible/module_utils/basic.py\", line 397, in _remove_values_conditions\nTypeError: Value of unknown type: <class 'pyvcloud.vcd.exceptions.UnauthorizedException'>, Status code: 401, <empty response body> (request id: 87a157ab-ca00-4f5a-95e0-e4f15a0b9bd5)\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}

PLAY RECAP **************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

with -vvv

The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py", line 327, in main
  File "/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py", line 131, in manage_states
  File "/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py", line 173, in create
  File "/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py", line 161, in get_org
  File "/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py", line 1562, in get_org_by_name
    orgs = self._get_wk_resource(_WellKnownEndpoint.ORG_LIST)
  File "/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py", line 1659, in _get_wk_resource
    return self.get_resource(self._get_wk_endpoint(wk_type))
  File "/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py", line 1452, in get_resource
    extra_headers=extra_headers)
  File "/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py", line 1169, in _do_request
    _objectify_response(response, objectify_results))
  File "/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py", line 1177, in _response_code_to_exception
    raise UnauthorizedException(sc, request_id, objectify_response)
pyvcloud.vcd.exceptions.UnauthorizedException: Status code: 401, <empty response body> (request id: e2e3c01b-98f8-4096-8649-bf2a0d489cf4)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/srvc_ansible/.ansible/tmp/ansible-tmp-1613711813.0736651-16647-144031486706802/AnsiballZ_vcd_org.py", line 102, in <module>
    _ansiballz_main()
  File "/home/srvc_ansible/.ansible/tmp/ansible-tmp-1613711813.0736651-16647-144031486706802/AnsiballZ_vcd_org.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/srvc_ansible/.ansible/tmp/ansible-tmp-1613711813.0736651-16647-144031486706802/AnsiballZ_vcd_org.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible.modules.vcd_org', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python3.6/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py", line 341, in <module>
  File "/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py", line 335, in main
  File "/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/module_utils/basic.py", line 2214, in fail_json
  File "/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/module_utils/basic.py", line 2187, in _return_formatted
  File "/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/module_utils/basic.py", line 420, in remove_values
  File "/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/module_utils/basic.py", line 397, in _remove_values_conditions
TypeError: Value of unknown type: <class 'pyvcloud.vcd.exceptions.UnauthorizedException'>, Status code: 401, <empty response body> (request id: e2e3c01b-98f8-4096-8649-bf2a0d489cf4)
fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py\", line 327, in main\n  File \"/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py\", line 131, in manage_states\n  File \"/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py\", line 173, in create\n  File \"/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py\", line 161, in get_org\n  File \"/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py\", line 1562, in get_org_by_name\n    orgs = self._get_wk_resource(_WellKnownEndpoint.ORG_LIST)\n  File \"/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py\", line 1659, in _get_wk_resource\n    return self.get_resource(self._get_wk_endpoint(wk_type))\n  File \"/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py\", line 1452, in get_resource\n    extra_headers=extra_headers)\n  File \"/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py\", line 1169, in _do_request\n    _objectify_response(response, objectify_results))\n  File \"/home/srvc_ansible/.local/lib/python3.6/site-packages/pyvcloud/vcd/client.py\", line 1177, in _response_code_to_exception\n    raise UnauthorizedException(sc, request_id, objectify_response)\npyvcloud.vcd.exceptions.UnauthorizedException: Status code: 401, <empty response body> (request id: e2e3c01b-98f8-4096-8649-bf2a0d489cf4)\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/home/srvc_ansible/.ansible/tmp/ansible-tmp-1613711813.0736651-16647-144031486706802/AnsiballZ_vcd_org.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/home/srvc_ansible/.ansible/tmp/ansible-tmp-1613711813.0736651-16647-144031486706802/AnsiballZ_vcd_org.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/srvc_ansible/.ansible/tmp/ansible-tmp-1613711813.0736651-16647-144031486706802/AnsiballZ_vcd_org.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible.modules.vcd_org', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib64/python3.6/runpy.py\", line 205, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.6/runpy.py\", line 96, in _run_module_code\n    mod_name, mod_spec, pkg_name, script_name)\n  File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py\", line 341, in <module>\n  File \"/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/modules/vcd_org.py\", line 335, in main\n  File \"/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/module_utils/basic.py\", line 2214, in fail_json\n  File \"/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/module_utils/basic.py\", line 2187, in _return_formatted\n  File \"/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/module_utils/basic.py\", line 420, in remove_values\n  File \"/tmp/ansible_vcd_org_payload_75ps83ev/ansible_vcd_org_payload.zip/ansible/module_utils/basic.py\", line 397, in _remove_values_conditions\nTypeError: Value of unknown type: <class 'pyvcloud.vcd.exceptions.UnauthorizedException'>, Status code: 401, <empty response body> (request id: e2e3c01b-98f8-4096-8649-bf2a0d489cf4)\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}

I tried using local modules mentioned in ansible.cfg or copying them in all the following locations:
- /home/srvc_ansible/.ansible/plugins/modules
- /usr/share/ansible/plugins/modules
- /usr/local/lib/python3.6/site-packages/ansible/modules

None of them works!

I even tried using the pre-defined roles , but same error.

**Comment by TheEric7** [comment]
Anyone? Any Help?

**Comment by mukultaneja** [comment]
@TheEric7, the logs are saying you are not authorized to perform operations on VCD. Please check the login credentials and share your env details.

**Comment by TheEric7** [comment]
@mukultaneja, Thanks for the reply.
I have admin access on the VCD and just double-checked the credentials, they are correct.
The VCD version is 10.1.0.15967236. What else do you need?

**Comment by mukultaneja** [comment]
Could you please try changing your API version to `34.0` once

**Comment by TheEric7** [comment]
It isn't the API version either.
But the good news is that I somehow managed to find the issue and fix it.

Apparently, Ansible was not able to make a connection to the VCD and I believe that's why you thought it might be a permission issue. In fact, it seems that Ansible was not able to get to the VCD at all.

So I found out that our DNS is resolving the VCD to a wrong IP address and once I fixed that by adding a line in **/etc/hosts** the issue was gone. I tried to troubleshoot the issue in multiple ways but the reason I couldn't find it earlier was because of the Error Description. When you put a dummy address as the host, it gives you a different error and I thought that if I had a connection issue I would get the same.

Anyways, thanks for your reply. Hope it helps the others if they have the same issue.
